File: backend/api/schemas/audiencia.py
"""
Audiencia API Schemas

SPRINT4-T4.1: Schemas for aggregated audience data endpoints
"""
from pydantic import BaseModel
from typing import List, Optional


class TopicAggregation(BaseModel):
    """Aggregated conversation topic"""
    topic: str
    count: int
    percentage: float
    quotes: List[str]  # Real examples (max 5)
    users: List[str]   # Usernames who mentioned it


class ObjectionAggregation(BaseModel):
    """Aggregated objection"""
    objection: str
    count: int
    percentage: float
    quotes: List[str]
    suggestion: str
    resolved_count: int = 0
    pending_count: int = 0


class CompetitionMention(BaseModel):
    """Competition mention"""
    competitor: str  # @username
    count: int
    sentiment: str   # "positivo", "neutral", "negativo"
    context: List[str]  # Phrases where mentioned
    suggestion: str


class TrendItem(BaseModel):
    """Emerging trend"""
    term: str
    count_this_week: int
    count_last_week: int
    growth_percentage: float
    quotes: List[str]


class ContentRequest(BaseModel):
    """Content requested by audience"""
    topic: str
    count: int
    questions: List[str]  # Specific questions
    suggestion: str


class ProductRequest(BaseModel):
    """Product requested that doesn't exist"""
    product_name: str
    count: int
    quotes: List[str]
    potential_revenue: float


class PerceptionItem(BaseModel):
    """What they think about you"""
    aspect: str  # "expertise", "precio", "atencion"
    positive_count: int
    negative_count: int
    quotes_positive: List[str]
    quotes_negative: List[str]


# Response models
class TopicsResponse(BaseModel):
    """Response for topics endpoint"""
    total_conversations: int
    topics: List[TopicAggregation]


class ObjectionsResponse(BaseModel):
    """Response for objections endpoint"""
    total_with_objections: int
    objections: List[ObjectionAggregation]


class CompetitionResponse(BaseModel):
    """Response for competition endpoint"""
    total_mentions: int
    competitors: List[CompetitionMention]


class TrendsResponse(BaseModel):
    """Response for trends endpoint"""
    period: str = "week"
    trends: List[TrendItem]


class ContentRequestsResponse(BaseModel):
    """Response for content requests endpoint"""
    total_requests: int
    requests: List[ContentRequest]


class ProductRequestsResponse(BaseModel):
    """Response for product requests endpoint"""
    total_requests: int
    products: List[ProductRequest]


class PerceptionResponse(BaseModel):
    """Response for perception endpoint"""
    total_analyzed: int
    perceptions: List[PerceptionItem]

File: frontend/src/services/api/audiencia.ts
import { apiFetch } from "./client";

export interface TopicAggregation {
  topic: string;
  count: number;
  percentage: number;
  quotes: string[];
  users: string[];
}

export interface ObjectionAggregation {
  objection: string;
  count: number;
  percentage: number;
  quotes: string[];
  suggestion: string;
  resolved_count: number;
  pending_count: number;
}

export interface CompetitionMention {
  competitor: string;
  count: number;
  sentiment: "positivo" | "neutral" | "negativo";
  context: string[];
  suggestion: string;
}

export interface TrendItem {
  term: string;
  count_this_week: number;
  count_last_week: number;
  growth_percentage: number;
  quotes: string[];
}

export interface ContentRequest {
  topic: string;
  count: number;
  questions: string[];
  suggestion: string;
}

export interface PerceptionItem {
  aspect: string;
  positive_count: number;
  negative_count: number;
  quotes_positive: string[];
  quotes_negative: string[];
}

export interface TopicsResponse {
  total_conversations: number;
  topics: TopicAggregation[];
}

export interface ObjectionsResponse {
  total_with_objections: number;
  objections: ObjectionAggregation[];
}

export interface CompetitionResponse {
  total_mentions: number;
  competitors: CompetitionMention[];
}

export interface TrendsResponse {
  period: string;
  trends: TrendItem[];
}

export interface ContentRequestsResponse {
  total_requests: number;
  requests: ContentRequest[];
}

export interface PerceptionResponse {
  total_analyzed: number;
  perceptions: PerceptionItem[];
}

export async function getAudienciaTopics(creatorId: string, limit: number = 10): Promise<TopicsResponse> {
  return apiFetch(`/audiencia/${creatorId}/topics?limit=${limit}`);
}

export async function getAudienciaPassions(creatorId: string, limit: number = 10): Promise<TopicsResponse> {
  return apiFetch(`/audiencia/${creatorId}/passions?limit=${limit}`);
}

export async function getAudienciaFrustrations(creatorId: string, limit: number = 10): Promise<ObjectionsResponse> {
  return apiFetch(`/audiencia/${creatorId}/frustrations?limit=${limit}`);
}

export async function getAudienciaCompetition(creatorId: string, limit: number = 10): Promise<CompetitionResponse> {
  return apiFetch(`/audiencia/${creatorId}/competition?limit=${limit}`);
}

export async function getAudienciaTrends(creatorId: string, limit: number = 10): Promise<TrendsResponse> {
  return apiFetch(`/audiencia/${creatorId}/trends?limit=${limit}`);
}

export async function getAudienciaContentRequests(creatorId: string, limit: number = 10): Promise<ContentRequestsResponse> {
  return apiFetch(`/audiencia/${creatorId}/content-requests?limit=${limit}`);
}

export async function getAudienciaPurchaseObjections(creatorId: string, limit: number = 10): Promise<ObjectionsResponse> {
  return apiFetch(`/audiencia/${creatorId}/purchase-objections?limit=${limit}`);
}

export async function getAudienciaPerception(creatorId: string): Promise<PerceptionResponse> {
  return apiFetch(`/audiencia/${creatorId}/perception`);
}
